dbms/src/Interpreters/Join.cpp

```diff
@@ -443,19 +443,26 @@ void Join::insertFromBlock(const Block & block, size_t stream_index)
                 const auto & join_partition = partitions[i];
                 auto partition_lock = join_partition->lockPartition();
                 partitions[i]->insertBlockForBuild(std::move(dispatch_blocks[i]));
-                if (join_partition->isSpill())
-                    blocks_to_spill = join_partition->trySpillBuildPartition(force_spill_partition_blocks, build_spill_config.max_cached_data_bytes_in_spiller, partition_lock);
+
+                if (hash_join_spill_context->isPartitionSpilled(i))
+                {
+                    if (hash_join_spill_context->updatePartitionRevocableMemory(force_spill_partition_blocks, i, join_partition->revocableBytes()))
+                        blocks_to_spill = join_partition->trySpillBuildPartition(partition_lock);
+                    else
+                        continue;
+                }
                 else
-                    stored_block = join_partition->getLastBuildBlock();
-                if (stored_block != nullptr)
                 {
                     size_t byte_before_insert = join_partition->getHashMapAndPoolByteCount();
-                    insertFromBlockInternal(stored_block, i);
+                    insertFromBlockInternal(join_partition->getLastBuildBlock(), i);
                     size_t byte_after_insert = join_partition->getHashMapAndPoolByteCount();
                     if likely (byte_after_insert > byte_before_insert)
                     {
                         join_partition->addMemoryUsage(byte_after_insert - byte_before_insert);
                     }
+                    auto ret = hash_join_spill_context->updatePartitionRevocableMemory(force_spill_partition_blocks, i, join_partition->revocableBytes());
```

Comment:
add some comment here, such as `because of the join_partition is not spilled, the result of updatePartitionRevocableMemory must be false now`?

Reply:
Done